lib/exabgp/rib/store.py

```diff
@@ -182,6 +182,14 @@ def insert_announced (self, change, force=False):
 				if self.cache and change_nlri_index not in self._seen.get(change.nlri.family(),{}):
 					return
 
+		if not force and self.cache:
+			announced = self._seen.get(change.nlri.family(),{})
+			if change_nlri_index in announced and change.nlri.action == OUT.ANNOUNCE:
+				old_change = announced[change_nlri_index]
+				# it is a duplicate route
+				if old_change.attributes.index() == change.attributes.index() and old_change.nlri.nexthop.index() == change.nlri.nexthop.index():
```

Comment:
So from my reading of the code starting at line 247 in rib/store.py it looks like we only ever set a change entry in announced that has an action of OUT.ANNOUNCE, if we process an OUT.WITHDRAW change we actually remove that nlri from announced. Wouldn't this mean that checkout the action of the old_change would always be OUT.ANNOUNCE?

```
if self.cache:
    announced = self._seen
    for change in changed:
        if change.nlri.action == OUT.ANNOUNCE:
            announced.setdefault(change.nlri.family(),{})[change.index()] = change
        else:
            family = change.nlri.family()
            if family in announced:
                announced[family].pop(change.index(),None)
```